keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -1023,26 +1025,146 @@ this enhancement.
 
 #### Enforcement Actions
 
-For phase 1, all violations implicitly result in a `deny` enforcement action.
+`ValidatingAdmissionPolicyBinding` resources may control how admission is
+enforced. This is performed using a single field. E.g.:
 
-For phase 2, we intend to support multiple enforcement actions.
+```yaml
+apiVersion: admissionregistration.k8s.io/v1alpha1
+kind: ValidatingAdmissionPolicyBinding
+...
+spec:
+  validationActions: [Warn, Audit] # required field
+```
 
-Use cases:
+The enum options will be:
+
+- `Deny`: Validation failures result in a denied request.
+- `Warn`: Validation failures are reported as warnings to the client. (xref: [Admisssion Webhook Warnings](https://kubernetes.io/blog/2020/09/03/warnings/#admission-webhooks))
+- `Audit`: Validation failures are published as audit events (see below Audit
+  Annotations section for details).
+- `Log`: The apiserver handling the admission request logs the validation failure.
+
+If, in the future, `ValidatingAdmissionPolicy` also introduces enforcement
+action fields, this `validationActions` field on
+`ValidatingAdmissionPolicyBinding` will specify the "maximum" enforcement (at
+most the enforcement will be what `validationActions` specifies).
 
-- Cluster admin would like to rollout a policies, sometimes in bulk, without
+Systems that need to aggregate validation failures may implement an [audit
+webhook
+backend](https://kubernetes.io/docs/tasks/debug/debug-cluster/audit/#webhook-backend). See
+below "Audit Events" for details.
+
+For singleton policies, the `validationActions` field will be set on the policy definition.
+
+Metrics will include validation action so that cluster administrators can monitor the
+validation failures of a binding before setting `validationActions` to `Deny`.
+
+This enables the following use cases:
```

Comment:
This looks like it would work well for rolling out a new policy, but it's less straightforward for updating an existing policy.

The safest way to roll out an update would probably be to:
1. Duplicate the policy (name it something like `-v2`), and make your edits to the copy
2. Copy all the policy bindings: set the `validationActions` on the copies to `['Warn', 'Audit']`
3. Verify the new policy works as expected
4. Update the new policy bindings to `Deny`
5. Delete the old bindings & policy.

That doesn't seem like the best user journey. One way this might be improved is if each expression could have a set of validation actions, then it could be simplified to just duplicate the modified expressions, edit the copy, and make them non-enforcing. That way you don't need to mess with all the bindings. I still don't love the copy-iterate-delete workflow though... WDYT?

Reply:
I agree update is a mess to update (change and expression, add an expression) right now.

Previously we had punted on this problem, in part because higher level constructs like policy engines can automate the copying of the underlying policy definitions and bindings. But having to copy the bindings, in particular, it really gross.

Would it be reasonable to split off this into a separate KEP update, maybe for 1.28? We could do something like:

- Add per-binding enforcement actions as proposed in this PR for 1.27 (bindings are not aware of individual expressions)
- In 1.28, look into per-definition enforcement actions with the aim of offering a paved road for users performing updates.

?

If this sounds good I'll add some notes about future plans.

Reply:
One argument in favor of having separate definitions as part of a rollout: it limits the possibility of a bad rollout affecting policy.

If, for some reason, "alpha definition code" (coining term) is integrated in such a way that the definition fails to ingest, that alpha code interferes with the efficacy of the production code.

It might be good to allow a binding to point to multiple definitions, which would allow updating the underlying code without the need to update the configured intent. This is similar to how updating a controller implicitly works today.

Reply:
I like the how a binding that can point to multiple definitions helps in workflows like this.

Reply:
(resolving for purposes of this PR. Discussion will continue when we tackle future work related to enforcement actions)